File: clients/android/app/src/main/java/local/visual/corepose/ClipRange.kt
package local.visual.corepose

object ClipRange {
    const val FRAME_STRIDE = 2

    fun collectSeeds(seeds: List<SeedMark>, seedBox: NormBox?): List<SeedMark> {
        if (seeds.isNotEmpty()) {
            return seeds.sortedBy { it.tMs }
        }
        if (seedBox != null) {
            return listOf(SeedMark(0.0, seedBox))
        }
        return emptyList()
    }

    fun seedAt(tMs: Double, seeds: List<SeedMark>, halfMs: Double): SeedMark? {
        var hit: SeedMark? = null
        var best = halfMs
        for (item in seeds) {
            val delta = kotlin.math.abs(item.tMs - tMs)
            if (delta <= best) {
                best = delta
                hit = item
            }
        }
        return hit
    }

    fun inPlayRange(tMs: Double, startMs: Double, endMs: Double): Boolean {
        return tMs in startMs..endMs
    }

    fun clampPlayheadMs(tMs: Long, startMs: Long, endMs: Long): Long {
        val hi = (endMs - 1L).coerceAtLeast(startMs)
        return tMs.coerceIn(startMs, hi)
    }

    fun pastPlayEnd(posMs: Long, endMs: Long): Boolean {
        return posMs >= endMs
    }

    fun shouldReseekToStart(posMs: Long, startMs: Long, seekPending: Boolean): Boolean {
        return !seekPending && posMs < startMs
    }

    fun strideMs(fps: Double): Long {
        val step = (1000.0 / maxOf(fps, 1.0)) * FRAME_STRIDE
        return step.toLong().coerceAtLeast(1L).let { rounded ->
            if (step - rounded >= 0.5) rounded + 1 else rounded
        }
    }

    fun halfMs(fps: Double): Double {
        return (1000.0 / maxOf(fps, 1.0)) * FRAME_STRIDE / 2.0
    }
}

File: clients/ios/VisualPose/ClipRange.swift
import Foundation

enum ClipRange {
    static let frameStride = 2

    static func collectSeeds(_ seeds: [SeedMark], seedBox: NormBox?) -> [SeedMark] {
        if !seeds.isEmpty { return seeds.sorted { $0.tMs < $1.tMs } }
        if let seedBox { return [SeedMark(tMs: 0, box: seedBox)] }
        return []
    }

    static func seedAt(tMs: Double, seeds: [SeedMark], halfMs: Double) -> SeedMark? {
        var hit: SeedMark?
        var best = halfMs
        for item in seeds {
            let delta = abs(item.tMs - tMs)
            if delta <= best {
                best = delta
                hit = item
            }
        }
        return hit
    }

    static func inPlayRange(tMs: Double, startMs: Double, endMs: Double) -> Bool {
        tMs >= startMs && tMs <= endMs
    }

    static func clampPlayheadMs(_ tMs: Int64, startMs: Int64, endMs: Int64) -> Int64 {
        let hi = max(startMs, endMs - 1)
        return min(max(tMs, startMs), hi)
    }

    static func pastPlayEnd(posMs: Int64, endMs: Int64) -> Bool {
        posMs >= endMs
    }

    static func shouldReseekToStart(posMs: Int64, startMs: Int64, seekPending: Bool) -> Bool {
        !seekPending && posMs < startMs
    }

    static func strideMs(fps: Double) -> Int64 {
        let step = (1000.0 / max(fps, 1)) * Double(frameStride)
        let truncated = Int64(step)
        let rounded = (step - Double(truncated) >= 0.5) ? truncated + 1 : truncated
        return max(1, rounded)
    }

    static func halfMs(fps: Double) -> Double {
        (1000.0 / max(fps, 1)) * Double(frameStride) / 2
    }
}
